src/utils/display-observer.js

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright 2020 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import {getServiceForDoc, registerServiceBuilderForDoc} from '../service';
+import {pushIfNotExist, removeItem} from './array';
+import {rethrowAsync} from '../log';
+
+const SERVICE_ID = 'DisplayObserver';
+
+const DOC_OBSERVER = 0;
+const VIEWPORT_OBSERVER = 1;
+
+/** @implements {Disposable} */
+export class DisplayObserver {
+  /**
+   * @param {!AmpDoc} ampdoc
+   */
+  constructor(ampdoc) {
+    const {win} = ampdoc;
+
+    /** @private @const {!IntersectionObserver} */
+    this.docObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, DOC_OBSERVER),
+      {
+        root: ampdoc.getBody(),
+        threshold: 0.001,
+      }
+    );
+
+    /** @private @const {!IntersectionObserver} */
+    this.viewportObserver_ = new win.IntersectionObserver(
+      (e) => this.observed_(e, VIEWPORT_OBSERVER),
+      {threshold: 0.001}
+    );
+
+    /** @private {boolean} */
+    this.isDocVisible_ = ampdoc.isVisible();
+
+    /** @private {?UnlistenDef} */
+    this.visibilityUnlisten_ = ampdoc.onVisibilityChanged(() => {
+      if (ampdoc.isVisible() !== this.isDocVisible_) {
+        this.isDocVisible_ = ampdoc.isVisible();
+        this.docVisibilityChanged_();
+      }
+    });
+
+    /** @private @const {!Map<!Element, !Array<function(boolean)>>} */
+    this.targetObserverMultimap_ = new Map();
```

Comment:
Given we have a `targetDisplayMap_`, we should make this more explicitly about callbacks:

```suggestion
    this.targetObserverCallbacks_ = new Map();
```